tools/envoy_headersplit/headersplit.py

```diff
@@ -0,0 +1,391 @@
+        the include statements into generated headers. Return string seems more convenient
+    """
+
+  cursor = translation_unit.cursor
+  for descendant in cursor.walk_preorder():
+    if descendant.location.file is not None and descendant.location.file.name == cursor.displayname:
+      filename = descendant.location.file.name
+      with open(filename, 'r') as source_file:
+        contents = source_file.read()
+      headers = contents[:descendant.extent.start.offset]
+      return headers
+
+  return ""
+
+
+def class_definitions(cursor: Cursor) -> List[Cursor]:
+  """
+    extracts all class definitions in the file pointed by cursor. (typical mocks.h)
+
+    Args:
+        cursor: cursor of parsing result of target souce code by libclang
+
+    Returns:
+        a list of cursor, each pointing to a class definition.
+
+    """
+  class_cursors = []
+  for descendant in cursor.walk_preorder():
+    # We don't want Cursors from files other than the one belonging to
+    # translation_unit, otherwise we get definitions for every file included
+    # when clang parsed the input file (i.e. if we don't limit descendant loction,
+    # it will check definitions from included headers and get class defns like std::string)
+    if descendant.location.file is None:
+      continue
+    if descendant.location.file.name != cursor.displayname:
+      continue
+
+    # check if descendant is pointing a class delaration block.
+    if descendant.kind != CursorKind.CLASS_DECL:
+      continue
+    if not descendant.is_definition():
+      continue
+
+    # check if this class is directly enclosed by a namespace.
+    # if it's not, then it's a nested class, our divided results
+    # should not contain nested classes.
+    if descendant.semantic_parent.kind != CursorKind.NAMESPACE:
+      continue
+
+    class_cursors.append(descendant)
+  return class_cursors
+
+
+def class_implementations(cursor: Cursor) -> List[Cursor]:
+  """
+    extracts all class implementation in the file pointed by cursor. (typical mocks.cc)
+
+    Args:
+        cursor: cursor of parsing result of target souce code by libclang
+
+    Returns:
+        a list of cursor, each pointing to a class implementation.
+
+    """
+  impl_cursors = []
+  for descendant in cursor.walk_preorder():
+    # We don't want Cursors from files other than the one belonging to
+    # translation_unit, otherwise we get definitions for every file included
+    # when clang parsed the input file (i.e. if we don't limit descendant loction,
+    # it will check definitions from included headers and get class defns like std::string)
+    if descendant.location.file is None:
+      continue
+    if descendant.location.file.name != cursor.displayname:
+      continue
+    if descendant.kind == CursorKind.NAMESPACE:
+      continue
+    if descendant.semantic_parent is not None and descendant.semantic_parent.kind == CursorKind.CLASS_DECL:
+      impl_cursors.append(descendant)
```

Comment:
This walk pattern is similar to the above method. Would it be reasonable to factor it out?